src/rule.rs

```diff
@@ -133,12 +153,44 @@ pub struct RuleSubvolume {
 }
 
 impl RuleSubvolume {
-    fn snapshot(&self) -> anyhow::Result<()> {
+    fn snapshot(&self, rule: &Rule) -> anyhow::Result<()> {
         let mut name = RandomName::new();
         for _ in 0..16 {
             let target_path = self.target_dir.join(name.as_str());
+            if !rule.pre_snapshot.is_empty() {
+                let mut cmd = std::process::Command::new("sh");
+                cmd.arg("-c").arg(format!("{} {} {}", rule.pre_snapshot, self.path.display(), target_path.display()));
+                let res = cmd.output();
+                match res {
+                    Ok(output) => {
+                        if !output.status.success() {
+                            let error_code = output.status.code();
+                            if let Some(code) = error_code {
+                                eprintln!("Pre Snapshot script failed, aborting snapshot, exit code: {}, error: {}", code, String::from_utf8_lossy(&output.stderr));
+                            } else {
+                                eprintln!("Pre Snapshot script failed, aborting snapshot, error: {}", String::from_utf8_lossy(&output.stderr));
+                            }
+                            return Err(anyhow::anyhow!("Pre snapshot script failed"));
```

Comment:
should we abort snapshotting because user's script failed?

Reply:
my usecase here is to use the pre snapshot to check / mount the external backup location,
local snapshots are useless for me as a backup thing if it is not kept externally,
using a pre action script to check things before letting action happen is a pattern I have seen and could be useful for other usecases.
With btrfs snapshots being so cheap, there isnt really much difference otherwise between pre and post, except that pre can prevent backup from happening if so desired.